File: packaging/aur/src/fpvtop-0.2.0/web/draw.js
const GRAPH_UP = [
  " ", "⢀", "⢠", "⢰", "⢸",
  "⡀", "⣀", "⣠", "⣰", "⣸",
  "⡄", "⣄", "⣤", "⣴", "⣼",
  "⡆", "⣆", "⣦", "⣶", "⣾",
  "⡇", "⣇", "⣧", "⣷", "⣿",
];

const GRAPH_DOWN = [
  " ", "⠈", "⠘", "⠸", "⢸",
  "⠁", "⠉", "⠙", "⠹", "⢹",
  "⠃", "⠋", "⠛", "⠻", "⢻",
  "⠇", "⠏", "⠟", "⠿", "⢿",
  "⡇", "⡏", "⡟", "⡿", "⣿",
];

const METER_CH = "■";

function clampv(v, lo, hi) {
  return v < lo ? lo : v > hi ? hi : v;
}

function span(color, text) {
  return '<span style="color:' + color + '">' + text + "</span>";
}

class Graph {
  constructor(width, height, gradient, invert = false, noZero = false) {
    this.width = width;
    this.height = height;
    this.gradient = gradient;
    this.invert = invert;
    this.noZero = noZero;
    this.current = true;
    this.last = 0;
    this.graphs = { true: [], false: [] };
    for (const key of [true, false]) {
      for (let r = 0; r < height; r++) this.graphs[key].push(Array(width).fill(" "));
    }
  }

  add(value) {
    value = clampv(Math.round(value), 0, 100);
    this.current = !this.current;
    for (const row of this.graphs[this.current]) if (row.length) row.shift();
    const syms = this.invert ? GRAPH_DOWN : GRAPH_UP;
    const mod = this.height === 1 ? 0.3 : 0.1;
    for (let horizon = 0; horizon < this.height; horizon++) {
      const curHigh = this.height > 1 ? Math.round((100 * (this.height - horizon)) / this.height) : 100;
      const curLow = this.height > 1 ? Math.round((100 * (this.height - horizon - 1)) / this.height) : 0;
      const result = [this.last, value].map((v, ai) => {
        const clampMin = this.noZero && horizon === this.height - 1 ? 1 : 0;
        if (v >= curHigh) return 4;
        if (v <= curLow) return clampMin;
        return clampv(Math.round(((v - curLow) * 4) / (curHigh - curLow) + mod), clampMin, 4);
      });
      const sym = syms[result[0] * 5 + result[1]];
      if (this.height === 1 && this.gradient) {
        const color = this.gradient[clampv(Math.max(this.last, value), 0, 100)];
        this.graphs[this.current][horizon].push(result[0] + result[1] === 0 ? " " : span(color, sym));
      } else {
        this.graphs[this.current][horizon].push(sym);
      }
    }
    this.last = value;
  }

  html() {
    const rows = [];
    for (let i = 1; i <= this.height; i++) {
      const row = this.graphs[this.current][this.invert ? this.height - i : i - 1].join("");
      if (this.height === 1) {
        rows.push(row);
      } else if (this.gradient) {
        const idx = this.invert ? Math.floor((i * 100) / this.height) : 100 - Math.floor(((i - 1) * 100) / this.height);
        rows.push(span(this.gradient[idx], row));
      } else {
        rows.push(row);
      }
    }
    return rows.join("\n");
  }
}

class MeterDraw {
  constructor(width, gradient) {
    this.width = width;
    this.gradient = gradient;
    this.cache = {};
  }

  html(value) {
    value = clampv(Math.round(value), 0, 100);
    if (this.cache[value]) return this.cache[value];
    let out = "";
    for (let i = 1; i <= this.width; i++) {
      const y = Math.round((i * 100) / this.width);
      if (value >= y) {
        out += span(this.gradient[y], METER_CH);
      } else {
        out += span(THEME.meter_bg, METER_CH.repeat(this.width + 1 - i));
        break;
      }
    }
    this.cache[value] = out;
    return out;
  }
}

function padLeft(text, width) {
  text = String(text);
  return text.length >= width ? text.slice(0, width) : " ".repeat(width - text.length) + text;
}

function padRight(text, width) {
  text = String(text);
  return text.length >= width ? text.slice(0, width) : text + " ".repeat(width - text.length);
}

function divLabel(label, value, cols) {
  const fill = cols - label.length - 1 - value.length - 3;
  return (
    span(THEME.div_line, "─") +
    span(THEME.title, label + ":") +
    span(THEME.div_line, "─".repeat(Math.max(1, fill))) +
    span(THEME.title, value.replace(/ /g, "─")) +
    span(THEME.div_line, "─")
  );
}

File: fpvtop/draw.py
from collections import deque

from . import symbols as Sym
from .terminal import Fx, Mv


def clamp(value, low, high):
    return low if value < low else high if value > high else value


class Graph:
    def __init__(self, width, height, gradient=None, invert=False, no_zero=False,
                 max_value=0, offset=0):
        self.width = width
        self.height = height
        self.gradient = gradient
        self.invert = invert
        self.no_zero = no_zero
        self.offset = offset
        self.max_value = 100 if (max_value == 0 and offset > 0) else max_value
        self.current = True
        self.last = 0
        self.out = ""
        self.data = deque(maxlen=width * 2 + 4)
        self.graphs = {
            True: [[" "] * width for _ in range(height)],
            False: [[" "] * width for _ in range(height)],
        }
        self._build_out()

    def _create(self, data, data_offset):
        mult = (len(data) - data_offset) > 1
        syms = Sym.GRAPH_DOWN if self.invert else Sym.GRAPH_UP
        mod = 0.3 if self.height == 1 else 0.1
        data_value = 0
        if mult and data_offset > 0:
            self.last = data[data_offset - 1]
            if self.max_value > 0:
                self.last = clamp((self.last + self.offset) * 100 // self.max_value, 0, 100)
        for i in range(data_offset, len(data)):
            if mult:
                self.current = not self.current
            if i < 0:
                data_value = 0
                self.last = 0
            else:
                data_value = data[i]
                if self.max_value > 0:
                    data_value = clamp((data_value + self.offset) * 100 // self.max_value, 0, 100)
            for horizon in range(self.height):
                cur_high = round(100.0 * (self.height - horizon) / self.height) if self.height > 1 else 100
                cur_low = round(100.0 * (self.height - (horizon + 1)) / self.height) if self.height > 1 else 0
                result = []
                for ai, value in enumerate((self.last, data_value)):
                    clamp_min = 1 if (self.no_zero and horizon == self.height - 1
                                      and not (mult and i == data_offset and ai == 0)) else 0
                    if value >= cur_high:
                        result.append(4)
                    elif value <= cur_low:
                        result.append(clamp_min)
                    else:
                        result.append(clamp(int(round((value - cur_low) * 4 / (cur_high - cur_low) + mod)),
                                            clamp_min, 4))
                if self.height == 1:
                    if result[0] + result[1] == 0:
                        self.graphs[self.current][horizon].append(" ")
                    else:
                        cell = ""
                        if self.gradient:
                            cell = self.gradient[clamp(max(self.last, data_value), 0, 100)]
                        self.graphs[self.current][horizon].append(cell + syms[result[0] * 5 + result[1]])
                else:
                    self.graphs[self.current][horizon].append(syms[result[0] * 5 + result[1]])
            if mult and i >= 0:
                self.last = data_value
        self.last = data_value
        self._build_out()

    def _build_out(self):
        if self.height == 1:
            row = self.graphs[self.current][0]
            self.out = "".join(row) + (Fx.reset if self.gradient else "")
            return
        parts = []
        for i in range(1, self.height + 1):
            if i > 1:
                parts.append(Mv.d(1) + Mv.l(self.width))
            if self.gradient:
                idx = i * 100 // self.height if self.invert else 100 - ((i - 1) * 100 // self.height)
                parts.append(self.gradient[idx])
            row = self.graphs[self.current][self.height - i] if self.invert else self.graphs[self.current][i - 1]
            parts.append("".join(row))
        parts.append(Fx.reset)
        self.out = "".join(parts)

    def add(self, value):
        self.data.append(int(value))
        self.current = not self.current
        for row in self.graphs[self.current]:
            if row:
                row.pop(0)
        self._create(self.data, len(self.data) - 1)
        return self.out

    def __call__(self):
        return self.out


class Meter:
    def __init__(self, width, gradient, bg_color, invert=False):
        self.width = width
        self.gradient = gradient
        self.bg_color = bg_color
        self.invert = invert
        self.cache = {}

    def __call__(self, value):
        value = clamp(int(value), 0, 100)
        if value in self.cache:
            return self.cache[value]
        out = []
        for i in range(1, self.width + 1):
            y = round(i * 100 / self.width)
            if value >= y:
                out.append(self.gradient[100 - y if self.invert else y] + Sym.METER)
            else:
                out.append(self.bg_color + Sym.METER * (self.width + 1 - i))
                break
        out.append(Fx.reset)
        result = "".join(out)
        self.cache[value] = result
        return result


def create_box(x, y, width, height, theme, line_color=None, fill=False,
               title="", title2="", num=0):
    if not line_color:
        line_color = theme.c("div_line")
    numbering = "" if num == 0 else theme.c("hi_fg") + Sym.SUPERSCRIPT[clamp(num, 0, 9)]
    out = [Fx.reset, line_color]
    for hpos in (y, y + height - 1):
        out.append(Mv.to(hpos, x) + Sym.H_LINE * (width - 1))
    for hpos in range(y + 1, y + height - 1):
        out.append(Mv.to(hpos, x) + Sym.V_LINE
                   + (" " * (width - 2) if fill else Mv.r(width - 2))
                   + Sym.V_LINE)
    out.append(Mv.to(y, x) + Sym.ROUND_LEFT_UP)
    out.append(Mv.to(y, x + width - 1) + Sym.ROUND_RIGHT_UP)
    out.append(Mv.to(y + height - 1, x) + Sym.ROUND_LEFT_DOWN)
    out.append(Mv.to(y + height - 1, x + width - 1) + Sym.ROUND_RIGHT_DOWN)
    if title:
        out.append(Mv.to(y, x + 2) + Sym.TITLE_LEFT + Fx.b + numbering
                   + theme.c("title") + title + Fx.ub + line_color + Sym.TITLE_RIGHT)
    if title2:
        out.append(Mv.to(y + height - 1, x + 2) + Sym.TITLE_LEFT_DOWN + Fx.b + numbering
                   + theme.c("title") + title2 + Fx.ub + line_color + Sym.TITLE_RIGHT_DOWN)
    out.append(Fx.reset + Mv.to(y + 1, x + 1))
    return "".join(out)


def embed_title(theme, line_color, text, bold=False):
    inner = (Fx.b if bold else "") + theme.c("title") + text + (Fx.ub if bold else "")
    return line_color + Sym.TITLE_LEFT + inner + line_color + Sym.TITLE_RIGHT


def embed_title_down(theme, line_color, text):
    return (line_color + Sym.TITLE_LEFT_DOWN + theme.c("title") + text
            + line_color + Sym.TITLE_RIGHT_DOWN)
